src/mono/mono/mini/mini-arm64.c

```diff
@@ -507,24 +507,39 @@ emit_imm (guint8 *code, int dreg, int imm)
 static guint8*
 emit_imm64 (guint8 *code, int dreg, guint64 imm)
 {
-	if (imm == 0) {
-		arm_movzx (code, dreg, 0, 0);
-	} else {
-		gboolean is_inited = FALSE;
+	// Determine if there is more advantage to using movn or movz for initialization.
+	int num0 = 0, num1 = 0;
+	for (int idx = 0; idx < 64; idx+=16) {
+		int w = (imm >> idx) & 0xffff;
+		if (w == 0)
+			num0++;
+		else if (w == 0xffff)
+			num1++;
+	}
 
-		for (int idx = 0; idx < 64; idx += 16) {
-			int w = (imm >> idx) & 0xffff;
+	gboolean is_negated = num1 > num0;
+	gboolean is_inited = FALSE;
 
-			if (w) {
-				if (is_inited) {
-					arm_movkx (code, dreg, w, idx);
-				} else {
-					arm_movzx (code, dreg, w, idx);
-					is_inited = TRUE;
-				}
-			}
+	for (int idx = 0; idx < 64; idx+=16) {
+		int w = (imm >> idx) & 0xffff;
+		if (!is_inited && is_negated && w != 0xffff) {
```

Comment:
Shouldn't it break out of the loop early ?

Reply:
I mean does this loop really has to do further iterations after it emitted the instructions ?

Reply:
The loop has to go over all the 16b words, it can emit anywhere from 1 to 4 instructions. E.g. `0x0001_0000_0000_0001` will emit:
```asm
movz x0, #0x1
; nothing at idx=16 and idx=32
movk x0, #0x1 lsl 48
```